File: packages/ts/src/operationResponse.ts
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import { ReallyMeCryptoError } from "./errors.js";
import {
  ensureByteArray,
  ensureIndependentByteArray,
} from "./validateBytes.js";
import { requireReallyMeWasmProvider } from "./wasmProvider.js";
import type { ReallyMeWasmProvider } from "./wasmProvider.js";

const MAX_CRYPTO_OPERATION_RESPONSE_BYTES = 1_048_608;

const readOperationResponse = (
  value: unknown,
  request: Uint8Array,
): Uint8Array => {
  if (!(value instanceof Uint8Array)) {
    throw new ReallyMeCryptoError("provider-failure");
  }
  ensureIndependentByteArray(value, [request]);
  if (
    value.length === 0 ||
    value.length > MAX_CRYPTO_OPERATION_RESPONSE_BYTES
  ) {
    value.fill(0);
    throw new ReallyMeCryptoError("provider-failure");
  }
  return value;
};

/**
 * Execute one generated binary protobuf request.
 *
 * The returned bytes are always a generated `CryptoOperationResponse` with a
 * typed `result` or typed `error` outcome.
 */
export const processOperationResponse = (
  requestBytes: Uint8Array,
): Uint8Array => {
  ensureByteArray(requestBytes);
  return processOperationResponseWithProvider(
    requireReallyMeWasmProvider(),
    requestBytes,
  );
};

export const processOperationResponseWithProvider = (
  provider: ReallyMeWasmProvider,
  requestBytes: Uint8Array,
): Uint8Array => {
  ensureByteArray(requestBytes);
  return readOperationResponse(
    provider.processOperationResponse(requestBytes),
    requestBytes,
  );
};

/**
 * Execute one permitted non-secret generated ProtoJSON request and return the
 * same binary `CryptoOperationResponse` used by `processOperationResponse`.
 * Secret-bearing selectors fail before JSON value deserialization.
 */
export const processOperationResponseJson = (
  requestJson: Uint8Array,
): Uint8Array => {
  ensureByteArray(requestJson);
  return processOperationResponseJsonWithProvider(
    requireReallyMeWasmProvider(),
    requestJson,
  );
};

export const processOperationResponseJsonWithProvider = (
  provider: ReallyMeWasmProvider,
  requestJson: Uint8Array,
): Uint8Array => {
  ensureByteArray(requestJson);
  return readOperationResponse(
    provider.processOperationResponseJson(requestJson),
    requestJson,
  );
};

File: packages/swift/Sources/ReallyMeCrypto/OperationResponse.swift
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

typealias ReallyMeCryptoOperationResponseFunction = @convention(c) (
    UnsafePointer<UInt8>?,
    Int,
    UnsafeMutablePointer<UInt8>?,
    Int,
    UnsafeMutablePointer<Int>?
) -> Int32

struct ReallyMeRustCAbiOperationResponseProcessor {
    // Mirrors MAX_CRYPTO_PROTO_MESSAGE_BYTES plus protobuf wrapper overhead.
    // Validate before allocation because the dylib path is caller-configurable.
    private static let maxProcessOutputLength = 1_048_608

    private let processOperationResponseFunction: ReallyMeCryptoOperationResponseFunction
    private let processOperationResponseJsonFunction: ReallyMeCryptoOperationResponseFunction

    init(library: ReallyMeRustCAbiLibrary) throws {
        processOperationResponseFunction = try library.loadFunction(
            "rm_crypto_process_operation_response",
            as: ReallyMeCryptoOperationResponseFunction.self
        )
        processOperationResponseJsonFunction = try library.loadFunction(
            "rm_crypto_process_operation_response_json",
            as: ReallyMeCryptoOperationResponseFunction.self
        )
    }

    init(
        processOperationResponseFunction: ReallyMeCryptoOperationResponseFunction,
        processOperationResponseJsonFunction: ReallyMeCryptoOperationResponseFunction
    ) {
        self.processOperationResponseFunction = processOperationResponseFunction
        self.processOperationResponseJsonFunction = processOperationResponseJsonFunction
    }

    func processOperationResponse(_ request: [UInt8]) throws -> [UInt8] {
        try process(request, with: processOperationResponseFunction)
    }

    func processOperationResponseJson(_ requestJson: [UInt8]) throws -> [UInt8] {
        try process(requestJson, with: processOperationResponseJsonFunction)
    }

    private func process(
        _ request: [UInt8],
        with function: ReallyMeCryptoOperationResponseFunction
    ) throws -> [UInt8] {
        var producedLength = 0
        var output = [UInt8](repeating: 0, count: Self.maxProcessOutputLength)
        let outputCapacity = output.count
        let status = request.withUnsafeBufferPointer { requestBuffer in
            output.withUnsafeMutableBufferPointer { outputBuffer in
                function(
                    requestBuffer.baseAddress,
                    request.count,
                    outputBuffer.baseAddress,
                    outputCapacity,
                    &producedLength
                )
            }
        }
        do {
            try ReallyMeRustCAbiStatus.throwIfError(status)
        } catch {
            ReallyMeCryptoMemory.bestEffortClear(&output)
            throw error
        }
        guard producedLength > 0, producedLength <= output.count else {
            ReallyMeCryptoMemory.bestEffortClear(&output)
            throw ReallyMeCryptoError.providerFailure
        }
        if producedLength < output.count {
            // The C ABI writes exactly producedLength bytes. Clear spare
            // capacity before truncating so Swift never retains stale bytes
            // from a future implementation that reuses backing storage.
            for index in producedLength..<output.count {
                output[index] = 0
            }
            output.removeSubrange(producedLength..<output.count)
        }
        return output
    }
}

extension ReallyMeCrypto {
    /// Executes one binary generated `CryptoOperationRequest`.
    ///
    /// The returned bytes are always a binary `CryptoOperationResponse` with a
    /// generated `CryptoOperationResult` or generated `CryptoError` outcome.
    public func processOperationResponse(_ request: [UInt8]) throws -> [UInt8] {
        guard let library = providers.rustCAbiLibrary else {
            throw ReallyMeCryptoError.providerFailure
        }
        return try ReallyMeRustCAbiOperationResponseProcessor(library: library).processOperationResponse(request)
    }

    /// Executes a permitted non-secret generated ProtoJSON request.
    ///
    /// JSON is request-only; secret-bearing selectors fail before value
    /// deserialization. Returned bytes use the same binary response as
    /// `processOperationResponse(_:)`.
    public func processOperationResponseJson(_ requestJson: [UInt8]) throws -> [UInt8] {
        guard let library = providers.rustCAbiLibrary else {
            throw ReallyMeCryptoError.providerFailure
        }
        return try ReallyMeRustCAbiOperationResponseProcessor(library: library).processOperationResponseJson(requestJson)
    }
}
